### PR TITLE
Add types to some components in the design package

### PR DESCRIPTION
**Pull request by bl-nero**
This change adds `@types/styled-components` and `@types/styled-system` packages and fixes all pending errors. Note that this doesn't mean that I added type information everywhere; that's merely a necessary first step for all of us (me included) to gradually increase the health of our codebase.

What we gain:
- No more unknown/defunct attributes in elements, no attribute name typos allowed.
- Strict typing of props in CSS interpolations.

What we're still missing:
- Support for typed theme values in HTML attributes. This would require upgrading the `styled-system` package itself, as the required typing changes were not backported to version 3 of `@types/styled-system`.

Non-trivial changes were verified with Storybook and (where it was easy) with a running app. Trivial changes were left to automated tests.

Some principles I observed:

- In places where I encountered attributes that could have impact on the UI if they were correct (i.e. they were typos or similar mistakes), I erred towards keeping existing UI look, with some small exceptions when I was sure what I was doing.
- Wherever possible, I made attributes (partivcularly boolean ones) optional to keep the burden on future authors to minimum.

This change should make it easier to work on the design system updates.

This PR is followed up by https://github.com/gravitational/teleport.e/pull/4233, and because these two are tightly coupled, there's no other way than to merge them as quick as possible one after another.

### CODE REVIEW

File: web/packages/teleterm/src/ui/TopBar/TopBarButton.tsx
Comment (gzdunek) [inline web/packages/teleterm/src/ui/TopBar/TopBarButton.tsx:21]: No need to export this.

File: web/packages/teleterm/src/ui/Vnet/VnetConnectionItem.tsx
Comment (gzdunek) [inline web/packages/teleterm/src/ui/Vnet/VnetConnectionItem.tsx:72]: What's the reason for extracting the interface?

File: web/packages/teleterm/src/ui/components/FieldInputs.tsx
Comment (gzdunek) [inline web/packages/teleterm/src/ui/components/FieldInputs.tsx:22]: Would be good to re-export the props from `'shared/components/FieldInput`.

File: web/packages/teleterm/src/ui/components/VirtualScroll/VirtualScroll.tsx
Comment (gzdunek) [inline web/packages/teleterm/src/ui/components/VirtualScroll/VirtualScroll.tsx:71]: I'd inline it.

File: web/packages/teleterm/src/ui/Tabs/TabItem.tsx
Comment (gzdunek) [inline web/packages/teleterm/src/ui/Tabs/TabItem.tsx:140]: No need to export + inline props?

File: web/packages/design/src/themed-components.ts
Comment (gzdunek) [inline web/packages/design/src/themed-components.ts:1]: Did we consider extending the default theme by using declaration merging?
https://styled-components.com/docs/api#create-a-declarations-file

We could put it in `teleport/web/@types/styled-components.d.ts`:
```ts
import 'styled-components';
import { Theme } from 'design/theme/themes/types';

declare module 'styled-components' {
  export interface DefaultTheme extends Theme {}
}
```

This would have two advantages:
* we will have the theme typed everywhere, not only in the `themed-components` module,
* we won't need this file at all (I think it may be confusing for others: where I should import `styled` from?).

I tried to do it and TS returned like 5 simple errors, so not much to address.

Comment (bl-nero) [reply web/packages/design/src/themed-components.ts:1]: Oh, so that's how you do it! I knew there was some alternative, but I somehow failed to find relevant docs. I'll definitely use this approach.

File: web/packages/teleterm/src/ui/Vnet/VnetConnectionItem.tsx
Comment (bl-nero) [reply web/packages/teleterm/src/ui/Vnet/VnetConnectionItem.tsx:72]: Because I thought I need to repeat it twice in the definition below. It turns out it's enough to do it once, in the `forwardRef` template parameters.

File: web/packages/shared/components/FileTransfer/FileTransferStateless/CommonElements.tsx
Comment (gzdunek) [inline web/packages/shared/components/FileTransfer/FileTransferStateless/CommonElements.tsx:25]: Now can be merged with the line 22.

**Review by gzdunek — APPROVED** [review]
Thanks for the work on this.
It's awesome that we will finally have types for the `design` components (and for the `theme`)!

File: web/packages/build/package.json
Comment (ravicious) [inline web/packages/build/package.json:57]: What's the compatibility between those packages and styled-components and styled-system? We're on 5.3.11 of styled-components, so I guess the types are valid for that version of that package. But I'm not sure about styled-system. We use 3.2.1 of styled-system, but this package here is on 5.1.22.

Comment (ravicious) [reply web/packages/build/package.json:57]: I don't think that's the way to proceed with this. External type definitions, such as those in the DefinitelyTyped repo, unlike type definitions directly in .ts files, can tell lies. As in, they are nothing more than statements saying "function `foo` returns `boolean`". TypeScript doesn't verify in any way if they're correct.

That's why it's so important that the definitions match the library version for which they were written for. styled-system went through [many changes between 3.2.1 and 5.1.5](https://github.com/styled-system/styled-system/blob/master/CHANGELOG.md), so I'd expect that the types in the newest package do not work with 3.2.1.

E.g., I see https://github.com/DefinitelyTyped/DefinitelyTyped/commit/5746d0876fca8f306d2319c0bf821511d550300b which updates the compatibility of those definitions to 4.1. The date of this commit corresponds to the release of 4.1.1 of `@types/styled-system` (`npm view @types/styled-system --json`).

I don't want us to run into problems down the line that stem from type definitions being incompatible with the library itself, because it'll be tricky to undo that. That's why I'd like to see a confirmation that 5.1.22 of `@types/styled-system` is indeed compatible with `@styled-system` 3.2.1. If there isn't any, the I think we should stick with the version of the definitions which was made with our version in mind.

---

When I change `@types/styled-system` to 3.2.2, which seems to correspond to our version of `styled-system`, then I do get around 25 extra errors. E.g., it complains about `<Flex border={1}>` being wrong here:

https://github.com/gravitational/teleport/blob/79b0385aa134b4e66a8821a0ab0e3d40cf1864ae/web/packages/shared/components/AccessRequests/NewRequest/RequestCheckout/SelectReviewers.tsx#L190-L193

And that seems to stem from the fact that in `@types/styled-system@3.2.2`, there's `BorderProps` which corresponds to the `border` prop. `Box`, which `Flex` inherits from `BordersProps`, `BorderRadiusProps`, `BorderColorProps`, but not `BorderProps`. Adding it there fixes the problem.

3.2.2 complains about `fontWeight="regular"`, since this is our custom defined font weight, but 3.2.2 has no idea about those. 5.1.22 appears to be fine with it and indeed, [the definitions seem to consider theme values](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d8da2b742fb2d05ad786d8afb42d1032c0fa2eff/types/styled-system/index.d.ts#L369-L378). But it turns out it's fine with it only because it accepts any string under that prop – `fontWeight="foo"` works just as fine. `FontWeightProps` accepts some extra type, but we don't provide any when using that type.

https://github.com/gravitational/teleport/blob/acb7510479d76a43ac207bbb435588321b98686f/web/packages/design/src/Text/Text.tsx#L39-L45

One could argue that it's at least not worse than it is right now, but then what do we gain from merging this PR? From what I see it's mostly better handling of custom-defined props on styled components such as in the example below. But are there any other mistakes that those types can catch?

https://github.com/gravitational/teleport/blob/acb7510479d76a43ac207bbb435588321b98686f/web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx#L95-L107

File: web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx
Comment (ravicious) [inline web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx:82]: @gzdunek This component here doesn't work quite well. ;f

@bl-nero Could you change this to `size={14}` to make it less broken?

File: web/packages/teleterm/src/ui/Tabs/TabItem.tsx
Comment (ravicious) [inline web/packages/teleterm/src/ui/Tabs/TabItem.tsx:144]: Yes, I believe it is used when there are no active tabs.

https://github.com/gravitational/teleport/blob/0fd5b0c29b7a3119644cf1856ddbaade96a59dfd/web/packages/teleterm/src/ui/Tabs/Tabs.tsx#L62-L64

https://github.com/gravitational/teleport/blob/0fd5b0c29b7a3119644cf1856ddbaade96a59dfd/web/packages/teleterm/src/ui/Tabs/TabItem.tsx#L55

https://github.com/gravitational/teleport/blob/0fd5b0c29b7a3119644cf1856ddbaade96a59dfd/web/packages/teleterm/src/ui/Tabs/TabItem.tsx#L77-L81

**Review by ravicious — APPROVED** [review]
What an effort! I looked through the changes in design and teleterm and stuff seems to look fine.

File: web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx
Comment (gzdunek) [reply web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx:82]: This request is so quick that it's hard to notice that something is broken hah

File: web/packages/design/src/LabelInput/LabelInput.tsx
Comment (bl-nero) [reply web/packages/design/src/LabelInput/LabelInput.tsx:26]: Yeah, our codebase isn't consistent about it either, so I just picked arbitrarily.

File: web/packages/design/src/Onboard/OnboardFooter.tsx
Comment (bl-nero) [reply web/packages/design/src/Onboard/OnboardFooter.tsx:30]: Oh, here's a tricky one. If you look at the `getTypography` function from `typography.ts`, you'll notice that it only makes sense if the argument is a string. It can't take literal theme objects, because it simply uses it unconditionally to fetch an item from the `theme.typography` map. Regardless of whether it should work this way, it means that this line's syntax effectively does nothing (indexing the `typography` map with an object yields an `undefined` value). "Fixing" this actually ruins the text (part of the line will be displayed in a different font), and it's been like that ever since the onboard footer component was introduced. So the _current looks_ of it is likely what it was supposed to look like in the first place, and as I'm keeping a rule of not modifying the looks until I'm 100% sure what I'm doing, I removed it as a no-op.

File: web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx
Comment (bl-nero) [reply web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx:82]: @ravicious @gzdunek It looks even worse with size={14}; I'm assuming you meant size={24} and a `line-height` of 0; please take a look and verify.

Comment (gzdunek) [reply web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx:82]: I think we avoid adding more props to this component and instead do this:

```
Subject: [PATCH] Fix the loading indicator
---
Index: web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx b/web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx
--- a/web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx	(revision c811b34f40977a0a1201f36c04708711b83d2129)
+++ b/web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx	(date 1718014860822)
@@ -17,7 +17,7 @@
  */
 
 import React, { useEffect, useState } from 'react';
-import { Box, ButtonPrimary, Flex, Indicator } from 'design';
+import { ButtonPrimary, Flex, Indicator } from 'design';
 import { fade } from 'design/theme/utils/colorManipulator';
 import styled from 'styled-components';
 
@@ -63,16 +63,14 @@
         width="100%"
         shouldDisplayIsLoading={shouldDisplayIsLoading}
       >
-        <Box mr="1">{`$`}</Box>
-        <span>{cliCommand}</span>
-        {shouldDisplayIsLoading && (
+        <span>$ {cliCommand}</span>
+        {!shouldDisplayIsLoading && (
           <Indicator
-            size={24}
+            color="inherit"
+            size="medium"
             delay="none"
             css={`
-              line-height: 0;
-              display: inline;
-              margin: auto 0 auto auto;
+              display: flex;
             `}
           />
         )}
@@ -99,6 +97,7 @@
   white-space: pre;
   word-break: break-all;
   font-size: 12px;
+  justify-content: space-between;
   color: ${props => {
     // always use light colors
     const { light } = props.theme.colors;

```
It makes it look like that:
image

File: web/packages/design/src/Onboard/OnboardFooter.tsx
Comment (ravicious) [reply web/packages/design/src/Onboard/OnboardFooter.tsx:30]: That makes total sense, thanks.

File: web/packages/teleterm/src/ui/Tabs/TabItem.tsx
Comment (ravicious) [reply web/packages/teleterm/src/ui/Tabs/TabItem.tsx:144]: Oh, I see. I just looked it up and it can be removed from `TabContent`. It used to be used for [selectively applying `cursor: pointer`](https://github.com/gravitational/teleport/blob/d4f5f2fce37ef41df478845f0d92bc30243eaacf/web/packages/teleterm/src/ui/Tabs/TabItem.tsx#L103-L105), but this code has since been removed. It still has to be passed to the `useTabDnD` hook though.

**Comment by bl-nero** [comment]
Heads up: although I have all required approvals, I'll hold off with merging this PR until I'm back from vacation on June 17th, since I don't feel it's the most responsible thing to do just before I depart.

File: web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx
Comment (ravicious) [reply web/packages/teleterm/src/ui/DocumentGateway/CliCommand.tsx:82]: This is how the new version looks like for me with scrollbars always on (since I use a mouse). But nvm, it's better than the previous version anyway and this call shouldn't take more than a few ms anyway.

https://github.com/gravitational/teleport/assets/27113/c04395fd-398b-49b3-8b12-b45eadc48ce7

**Comment by ravicious** [comment]
@bl-nero You can merge it and I'll take care of any follow ups if they're needed when you're away. It's better to do this sooner than later I think and it'll be less work for you when you get back.

**Comment by ravicious** [comment]
A couple of days ago I was actually considering if we should proceed with it now. We didn't manage to merge it before the test plan has started, so that means we'd be less confident in merging and specifically backporting it, and then it'll just cause further rift with the rest of the codebase, etc. So it all just seemed like bad timing.

But after all these are just types, so I think for the most part we should be fine.

File: web/packages/teleport/src/Discover/Shared/Mark.ts
Comment (ravicious) [inline web/packages/teleport/src/Discover/Shared/Mark.ts:21]: `light` doesn't seem to be used anywhere. But I guess we can keep it if we just want to avoid breaking anything.

**Review by ravicious — APPROVED** [review]
I went through shared and teleport changes just to be double sure and everything seems to be fine.

**Pull request merged**
